File: flappy_goud/Movement.cs
public class Movement
{
    public float Velocity { get; set; }
    public float Rotation { get; private set; }

    private readonly float gravity;
    private readonly float jumpStrength;
    private const float RotationSmoothing = 0.03f; // Adjusts how fast the bird's rotation responds

    private float jumpCooldownTimer = 0f;

    public Movement(float gravity, float jumpStrength)
    {
        this.gravity = gravity;
        this.jumpStrength = jumpStrength;
        Velocity = 0;
        Rotation = 0;
    }

    public void ApplyGravity(float deltaTime)
    {
        Velocity += gravity * deltaTime * GameConstants.TargetFPS;
        jumpCooldownTimer -= Math.Max(0, deltaTime);
    }

    public void TryJump(float deltaTime)
    {
        if (jumpCooldownTimer <= 0)
        {
            Jump();
            jumpCooldownTimer = GameConstants.JumpCooldown; // Reset the cooldown timer after jumping
        }
    }

    private void Jump()
    {
        Velocity = 0; // Reset velocity to 0 before applying gravity

        Velocity = jumpStrength * GameConstants.TargetFPS;
    }

    public void UpdatePosition(ref float positionY, float deltaTime)
    {
        positionY += Velocity * deltaTime;

        // Gradually update rotation based on velocity, with smoothing
        float targetRotation = Math.Clamp(Velocity * 3, -45, 45); // Controls max rotation angle based on velocity
        Rotation += (targetRotation - Rotation) * RotationSmoothing;
    }
}

File: examples/swift/flappy_bird/Sources/FlappyBird/Movement.swift
// Movement.swift

import Foundation

final class Movement {
    var velocity: Float = 0
    private(set) var rotation: Float = 0

    private let gravity: Float
    private let jumpStrength: Float
    private let rotationSmoothing: Float = 0.03
    private var jumpCooldownTimer: Float = 0

    init(gravity: Float, jumpStrength: Float) {
        self.gravity = gravity
        self.jumpStrength = jumpStrength
    }

    func applyGravity(deltaTime: Float) {
        velocity += gravity * deltaTime * Float(GameConstants.targetFPS)
        jumpCooldownTimer -= max(0, deltaTime)
    }

    func tryJump(deltaTime: Float) -> Bool {
        if jumpCooldownTimer <= 0 {
            jump()
            jumpCooldownTimer = GameConstants.jumpCooldown
            return true
        }
        return false
    }

    private func jump() {
        velocity = 0
        velocity = jumpStrength * Float(GameConstants.targetFPS)
    }

    func updatePosition(positionY: inout Float, deltaTime: Float) {
        positionY += velocity * deltaTime

        let targetRotation = min(max(velocity * 3, -45), 45)
        rotation += (targetRotation - rotation) * rotationSmoothing
    }
}
